<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Long Read Error Corrector Analyser</title>
    <link href="lib/css/handsontable.full.min.css" rel="stylesheet" type="text/css" />
    <link href="lib/css/style.css" rel="stylesheet" type="text/css" />
    <link href="lib/css/iziModal.min.css" rel="stylesheet" type="text/css" />

    <script src="lib/js/jquery.min.js"></script>
    <script src="lib/js/jquery-ui.min.js"></script>
    <script src="lib/js/jquery.blockUI.js"></script>
    <script src="lib/js/handsontable.full.min.js"></script>
    <script src="lib/js/js.js"></script>
    <script src="lib/js/plotly-latest.min.js"></script>
    <script src="lib/js/iziModal.min.js"></script>
    <script src="lib/js/anchorific.min.js"></script>



    <script>
        var plotInfo = {}; //will store all the data to be shown on clicking

        $(function(){ // on dom ready
            //fill the main subtables
            fillMainSubtable('LR_EC_read_stats_table', <statProfiler.getReadStatsAsJSArrayForHOT()>, <tools>);
            fillMainSubtable('LR_EC_base_stats_table', <statProfiler.getBaseStatsAsJSArrayForHOT()>, <tools>);
            fillMainSubtable('LR_EC_error_stats_table', <statProfiler.getErrorStatsAsJSArrayForHOT()>, <tools>);
            fillMainSubtable('LR_EC_annotation_stats_table', <statProfiler.getAnnotationStatsAsJSArrayForHOT()>, <tools>);

            //register the event when changing the format of the units in the main subtables
            infixes=["read", "base", "error"]
        	infixes.forEach(function(infix){
        		$("input[id^=LR_EC_"+infix+"_stats_table_number_format]").change(function(elem){
        			tableId = elem.target.name.replace("_number_format", "")
            		HOTs[tableId].render()
            	});
        	})

        	//register the event when changing which plot to show in Corretion collapsing paralogous gene families plots
        	$("input[id^=Collapsing_gene_families_show_option]").change(function(elem){
                //hide all plots first
                $("div[id^=htmlScatterPlotSizeParalogFamilies_]").css("display", "none")

                //show the plot we want
                divId = "htmlScatterPlotSizeParalogFamilies_" + elem.target.value
                $("#"+divId).css("display", "block")
            });

            //show the first Corretion collapsing paralogous gene families plot (the one checked by default)
            $("input[id^=Collapsing_gene_families_show_option]").each(function(i, elem){
            	if (elem.checked) {
            		divId = "htmlScatterPlotSizeParalogFamilies_" + elem.value
            		$("#"+divId).css("display", "block")
            	}
            })


            //hides all plots that have hideOnLoad attribute
            $("div").each(function(i, elem){
                if (attrExists(elem, "hideOnLoad"))
                    $(elem).css("display", "none")

            })

            //add the listener to the checkboxes which have the attribute percentageSwitcher
            $("input").each(function(i, elem){
                if (attrExists(elem, "percentageSwitcher")) {
                    $(elem).change(function(elem){
                        //get which element to change
                        idPrefix = elem.target.value

                        //hide all plots with the idPrefix first
                        $("div[id^="+idPrefix+"]").css("display", "none")

                        //show the plot we want
                        divId = idPrefix + (elem.currentTarget.checked ? "Percentage" : "Scalar")
                        $("#"+divId).css("display", "block")
                    })
                }
            });

            //fill the gene and transcript tables
            <JSCommentOutSimpleReport>
            fillFeatureTable('Gene', 'LR_EC_gene_table', <geneProfiler.geneProfileToJSArrayForHOT()>, <tools>);
            fillFeatureTable('Transcript', 'LR_EC_transcript_table', <geneProfiler.transcriptProfileToJSArrayForHOT()>, <tools>);
            <JSUncommentOutSimpleReport>

            //anchorify it
            $('body').anchorific();

            //all set - unblock the UI()
            unblockUI();
        })


    </script>
</head>
<body>

<script>
//let's first block the UI and draw everything
blockUI();
</script>
<!-- Div for the main table -->
<h1>Main stats</h1>

<h2>Read stats (compilation from AlignQC reports)</h2>

<p>Transform numbers in:</p>
<input type="radio" name="LR_EC_read_stats_table_number_format" id="LR_EC_read_stats_table_number_format_M" value="M"> Millions (M) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<input type="radio" name="LR_EC_read_stats_table_number_format" id="LR_EC_read_stats_table_number_format_K" value="K" checked> Kilos (K) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<input type="radio" name="LR_EC_read_stats_table_number_format" id="LR_EC_read_stats_table_number_format_N" value="N"> None<br>
<div id="LR_EC_read_stats_table"></div>
*Notes (from <a href="https://www.ncbi.nlm.nih.gov/pubmed/28868132" target="_blank">AlignQC publication</a>):<br/>
If two or more alignment paths are reasonably spaced across the read, and can together generate a longer alignment, they will be combined and classified as:<br/>
(a) a gapped alignment of a gene if paths occur within close proximity to each other on the same strand (GAPPED_ALIGN_READS);<br/>
(b) a trans-chimeric alignment if paths occur on different loci (TRANSCHIMERA_ALIGN_READS and CHIMERA_ALIGN_READS);<br/>
(c) a self-chimeric alignment if paths align to an overlapping genomic position (SELFCHIMERA_ALIGN_READS and CHIMERA_ALIGN_READS);<br/>
(d) otherwise, the read is defined as a single alignment (SINGLE_ALIGN_READS).<br/>

<table>
    <tr>
        <td><html_TOTAL_READS_Plot></td>
        <td><html_MEAN_LENGTH_Plot></td>
    </tr>
    <tr>
        <td><html_ALIGNED_READS_Plot></td>
        <td><html_UNALIGNED_READS_Plot></td>
    </tr>
    <tr>
        <td><html_SINGLE_ALIGN_READS_Plot></td>
        <td><html_GAPPED_ALIGN_READS_Plot></td>
    </tr>

    <tr>
        <td><html_CHIMERA_ALIGN_READS_Plot></td>
        <td><html_TRANSCHIMERA_ALIGN_READS_Plot></td>
    </tr>

    <tr>
        <td><html_SELFCHIMERA_ALIGN_READS_Plot></td>
        <td>&nbsp;</td>
    </tr>

</table>


<h2>Base stats (compilation from AlignQC reports)</h2>
<p>Transform numbers in:</p>
<input type="radio" name="LR_EC_base_stats_table_number_format" id="LR_EC_read_stats_table_number_format_M" value="M" checked> Millions (M) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<input type="radio" name="LR_EC_base_stats_table_number_format" id="LR_EC_read_stats_table_number_format_K" value="K"> Kilos (K) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<input type="radio" name="LR_EC_base_stats_table_number_format" id="LR_EC_read_stats_table_number_format_N" value="N"> None<br>
<div id="LR_EC_base_stats_table"></div>
*Notes (from <a href="https://www.ncbi.nlm.nih.gov/pubmed/28868132" target="_blank">AlignQC publication</a>):<br/>
If two or more alignment paths are reasonably spaced across the read, and can together generate a longer alignment, they will be combined and classified as:<br/>
(a) a gapped alignment of a gene if paths occur within close proximity to each other on the same strand (GAPPED_ALIGN_BASES);<br/>
(b) a trans-chimeric alignment if paths occur on different loci (TRANSCHIMERA_ALIGN_BASES and CHIMERA_ALIGN_BASES);<br/>
(c) a self-chimeric alignment if paths align to an overlapping genomic position (SELFCHIMERA_ALIGN_BASES and CHIMERA_ALIGN_BASES);<br/>
(d) otherwise, the read is defined as a single alignment (SINGLE_ALIGN_BASES).<br/>

<table>
    <tr>
        <td><html_TOTAL_BASES_Plot></td>
        <td><html_ALIGNED_BASES_Plot></td>
    </tr>
    <tr>
        <td><html_UNALIGNED_BASES_Plot></td>
        <td><html_SINGLE_ALIGN_BASES_Plot></td>
    </tr>
    <tr>
        <td><html_GAPPED_ALIGN_BASES_Plot></td>
        <td><html_CHIMERA_ALIGN_BASES_Plot></td>
    </tr>

    <tr>
        <td><html_TRANSCHIMERA_ALIGN_BASES_Plot></td>
        <td><html_SELFCHIMERA_ALIGN_BASES_Plot></td>
    </tr>
</table>


<h2>Error stats (compilation from AlignQC reports)</h2>
<p style="color: red">Note: these error stats are computed from 1 million bases of random reads, so you can expect some small differences <br/>
    in this metric between different runs.</p>
<div id="LR_EC_error_stats_table"></div>


<table>
    <tr>
        <td><html_ANY_ERROR_Plot></td>
        <td><html_MISMATCHES_Plot></td>
    </tr>
    <tr>
        <td><html_ANY_DELETION_Plot></td>
        <td><html_ANY_INSERTION_Plot></td>
    </tr>
    <tr>
        <td><html_COMPLETE_DELETION_Plot></td>
        <td><html_HOMOPOLYMER_DELETION_Plot></td>
    </tr>
    <tr>
        <td><html_COMPLETE_INSERTION_Plot></td>
        <td><html_HOMOPOLYMER_INSERTION_Plot></td>
    </tr>
</table>

<h2>Annotation stats (compilation from AlignQC reports)</h2>
<div id="LR_EC_annotation_stats_table"></div>
<table>
    <tr>
        <td><html_GENES_DETECTED_ANY_MATCH_Plot></td>
        <td><html_GENES_DETECTED_FULL_MATCH_Plot></td>
    </tr>
    <tr>
        <td><html_TRANSCRIPTS_DETECTED_ANY_MATCH_Plot></td>
        <td><html_TRANSCRIPTS_DETECTED_FULL_MATCH_Plot></td>
    </tr>
</table>

<br/><br/>
<h1>Does error-correction perturb long reads connectivity, losing transcript structure and exon coupling?</h1>
<p>If error correctors shorten a lot the long reads, their main advantages, which are describing the complete transcript structure and distant exon coupling, are lost.</p>
<p>It is important to keep the length of the long reads, since short reads cannot describe transcript structure and distant exon coupling.</p>


<input type="checkbox" value="htmlNbOfIdentifiedExonsLinePlot" percentageSwitcher> Show in %<br/>
<div id="htmlNbOfIdentifiedExonsLinePlotScalar">
<htmlNbOfIdentifiedExonsLinePlotScalar>
</div>
<div id="htmlNbOfIdentifiedExonsLinePlotPercentage" hideOnLoad>
<htmlNbOfIdentifiedExonsLinePlotPercentage>
</div>


<input type="checkbox" value="htmlHighestNbOfConsecutiveExonsLinePlot" percentageSwitcher> Show in %<br/>
<div id="htmlHighestNbOfConsecutiveExonsLinePlotScalar">
<htmlHighestNbOfConsecutiveExonsLinePlotScalar>
</div>
<div id="htmlHighestNbOfConsecutiveExonsLinePlotPercentage" hideOnLoad>
<htmlHighestNbOfConsecutiveExonsLinePlotPercentage>
</div>


<input type="checkbox" value="htmlFullPartialReadsPlot" percentageSwitcher> Show in %<br/>
<div id="htmlFullPartialReadsPlotScalar">
<htmlFullPartialReadsPlotScalar>
</div>
<div id="htmlFullPartialReadsPlotPercentage" hideOnLoad>
<htmlFullPartialReadsPlotPercentage>
</div>


<br/><br/>
<h1>Does error-correction perturb reads’ length?</h1>
<p>The points inside the filled blue area mean that the error corrector did not over-split the reads.</p>
<p>Beware of the points way over the filled black area.</p>

<htmlTotalReadsCuttingPlot>
<htmlAlignedReadsCuttingPlot>
<htmlUnalignedReadsCuttingPlot>


<h1>Does error-correction perturb the number of reads mapping to the genes and transcripts?</h1>
<p>x-axis represents the number of reads mapping to the genes/transcripts before the correction, and the y-axis, after correction.</p>
<h2>Gene level</h2>
<htmlScatterPlotCoverageOfGenes>
<h2>Transcript level</h2>
<htmlScatterPlotCoverageOfIsoforms>

<br/><br/>
<h1>Corretion collapsing paralogous gene families plots</h1>
To know how we computed the paralogous gene families, see: <a href="https://gitlab.com/leoisl/LR_EC_analyser/blob/master/GettingParalogs.txt" target="_blank">https://gitlab.com/leoisl/LR_EC_analyser/blob/master/GettingParalogs.txt</a><br/>
However, here we just take into consideration paralogous gene families with cardinality 2+.<br/>
This is because the large majority of gene families has cardinality 1 (see <a href="#paralogous-gene-families-size-distribution-">Paralogous gene families size distribution</a>) and such families are not interesting here, complicating the analysis.<br/>
Given a tool, the size of a <b>expressed</b> paralogous gene family is the number of genes expressed in a given gene family for that tool.<br/>
For example, if a gene family has 10 members, and 8 are expressed in the raw data, and 2 are expressed after the correction,<br/>
we can say that the size of this gene family before the correction is 8, and after, 2.<br/>
We can then conclude that the correction shrinked the paralogous gene family, probably due to collapsing genes.<br/>


<h2>General view of paralog families change per tool:</h2>
Here we compare the size of the gene families before and after correction in general terms.<br/>
<p style="color: red">Click on a bar to retrieve all gene families belonging to it!</p>
<htmlGeneralViewParalogFamilies>

<h2>Detailed view of paralog families change per tool:</h2>
This scatterplot shows the sizes of the paralogous gene families before and after correction, detailing what was shown in the previous plot.<br/>
The color code of a data point is the # of gene families on that data point.<br/>

<!-- TODO: not sure if this is still needed...
<input type="radio" name="Collapsing_gene_families_show_option" id="Collapsing_gene_families_show_option_ALL" value="A"> Show all paralogous gene families<br/>
<input type="radio" name="Collapsing_gene_families_show_option" id="Collapsing_gene_families_show_option_NO_DIAG" value="ND" checked> Hide paralogous genes families that were unchanged (diagonal line)<br/>
<input type="radio" name="Collapsing_gene_families_show_option" id="Collapsing_gene_families_show_option_NO_DIAG_NO_ZERO" value="NDN0"> Hide paralogous genes families that were unchanged (diagonal line) and keep only the ones expressed both before and after correction (hide gene families such that x=0 or y=0)<br/>

<div id="htmlScatterPlotSizeParalogFamilies_A" style="display: none;">
<htmlScatterPlotSizeParalogFamilies TODO: IF ADD BACK, PLEASE REMOVE THIS>
</div>

<div id="htmlScatterPlotSizeParalogFamilies_ND" style="display: none;">
<htmlScatterPlotSizeParalogFamiliesExcluingUnchanged TODO: IF ADD BACK, PLEASE REMOVE THIS>
</div>

<div id="htmlScatterPlotSizeParalogFamilies_NDN0" style="display: none;">
<htmlScatterPlotSizeParalogFamiliesExcluingUnchangedCommonGenes TODO: IF ADD BACK, PLEASE REMOVE THIS>
</div>
-->
<htmlScatterPlotSizeParalogFamilies>

<br/><br/>
<h1>Correction towards major isoform plots</h1>

<h2>Difference on the number of isoforms of a gene before and after correction - Gene scale</h2>
The genes taken into account here must be expressed in BOTH the raw reads AND the tool.<br/>
Negative values: isoforms of a gene were lost - it could be that they were just removed by the correction, or after correction all reads mapped to another gene.<br/>
Positive values: isoforms of a gene were gained.<br/>
<p style="color: red">Click on a bar to retrieve all genes belonging to it!</p>
<htmlDifferenceOnTheNumberOfIsoformsPlotIntersection>

<h2>Difference on the number of isoforms of a gene before and after correction - Gene-family scale</h2>
The gene <b>families</b> taken into account here must be expressed in BOTH the raw reads AND the tool.<br/>
Negative values: isoforms of a gene-family were lost - it could be that they were just removed by the correction, or after correction all reads mapped to another gene-family.<br/>
Positive values: isoforms of a gene-family were gained.<br/>
In comparison with the previous plot, it allows to check if paralogs "exchange" their isoforms after error-correction<br/>
Remember that we just take into consideration paralogous gene families with cardinality 2+.<br/>
<p style="color: red">Click on a bar to retrieve all genes belonging to it!</p>
<htmlDifferenceOnTheNumberOfIsoformsGeneFamilyPlotIntersection>

<h2>Lost transcripts in genes with Splicing Complexity 2+</h2>
The genes taken into account here must be expressed in BOTH the raw AND the tool taken into consideration.<br/>
Otherwise we could account for lost genes, and these would inflate the high relative transcript coverage bins (which is not what we want, since here we are investigating isoforms that were lost, not genes, probably corrected towards the main isoform)<br/>
The relative coverage or expression of a transcript is the number of reads mapping to the transcript divided by the number of reads mapping to its gene.<br/>

<h3>Non-normalized plot: </h3>
<p style="color: red">Click on a bar to retrieve all transcripts belonging to it!</p>
<htmlLostTranscriptInGenesWSP2Plot>

<h3>Normalized plot: </h3>
<p>Each bin is normalized with respect to the number of isoforms in bin (which is the number of isoforms with the given Relative transcript coverage in the raw reads).</p>
<p style="color: red">Click on a bar to retrieve all transcripts belonging to it!</p>
<htmlLostTranscriptInGenesWSP2NormalizedPlot>

<h2>Distribution of the difference of relative expression in the transcripts</h2>
This box and whisker plot represents the distribution of the difference of relative expression in the transcripts.<br/>
It was done by taking, for each tool, each transcript that is expressed either in the raw reads or in the tool, and computing the absolute difference between the relative expression of the transcript before and after correction.<br/>
The relative coverage or expression of a transcript is the number of reads mapping to the transcript divided by the number of reads mapping to its gene.<br/>
<htmlDifferencesInRelativeExpressionsBoxPlot>

<h2>Coverage of the main isoform before and after correction</h2>
The main isoform of a gene is the isoform with the highest expression in the raw dataset.<br/>
This plot shows how the coverage of the main isoform changes after the correction of each tool.<br/>
<htmlScatterPlotCoverageOfMainIsoforms>


<h1>Splice Sites Plots</h1>
<h2>Correct and Incorrect Splice Sites BarPlot (compilation from AlignQC reports)</h2>
<input type="checkbox" value="htmlCorrectIncorrectSSPlot" percentageSwitcher> Show in %<br/>
<div id="htmlCorrectIncorrectSSPlotScalar">
<htmlCorrectIncorrectSSPlotScalar>
</div>
<div id="htmlCorrectIncorrectSSPlotPercentage" hideOnLoad>
<htmlCorrectIncorrectSSPlotPercentage>
</div>

<h2>Detailed Incorrect Splice Sites BarPlots (processed from AlignQC reports)</h2>
<p><b>Incorrect SSs Near True SSs</b> is the number of incorrect Splice Sites near (at most 2 bps) away from the nearest true Splice Site.<br/>
    Such small errors could be due to splice alignment problems or small correction problems.<br/>
    Care must be taken when considering these small variations as true new splicing junctions.</p>
<p><b>Incorrect SSs with distance Multiple of 3</b> is the number of incorrect Splice Sites that are at a distance multiple of 3 of the nearest true Splice Site.<br/>
    A high value here might suggest many new splice junctions, but not always.<br/>
    For example, a new alternative acceptor site with a different ORF from the annotated ORF will cause the donor site to have a distance not multiple of 3 from the nearest true Splice Site.<br/>
    Note however that mutations changing the ORFs are uncommon.
</p>
<p><b>Incorrect SSs Far From True SSs</b> is the number of incorrect Splice Sites that are not near (more than 2 bps) away from the nearest true Splice Site and are at a distance that is NOT multiple of 3 of the nearest true Splice Site.</p>
<input type="checkbox" value="htmlDetailedIncorrectSSPlot" percentageSwitcher> Show in %<br/>
<div id="htmlDetailedIncorrectSSPlotScalar">
<htmlDetailedIncorrectSSPlotScalar>
</div>
<div id="htmlDetailedIncorrectSSPlotPercentage" hideOnLoad>
<htmlDetailedIncorrectSSPlotPercentage>
</div>

<h2>Incorrect Splice Site Distribution Plots (compilation from AlignQC reports)</h2>
<input type="checkbox" value="htmlSpliceSitesDistributionSSPlot" percentageSwitcher> Show in %
<div id="htmlSpliceSitesDistributionSSPlotScalar">
<htmlSpliceSitesDistributionSSPlotScalar>
</div>
<div id="htmlSpliceSitesDistributionSSPlotPercentage" hideOnLoad>
<htmlSpliceSitesDistributionSSPlotPercentage>
</div>

<HTMLCommentOutSimpleReport>
<br/><br/>
<h1>Gene read alignment viewer</h1>
Delta value is the largest difference between the # of reads mapping to a gene in a tool and in raw reads<br/>
<span style="color: red;">BE SURE TO DO THIS IF YOU WANT TO SEE THE IGV PLOTS: <a href="https://gitlab.com/leoisl/LR_EC_analyser#viewing-igv-plots-gene-stats-and-transcript-stats" target="_blank">https://gitlab.com/leoisl/LR_EC_analyser#viewing-igv-plots-gene-stats-and-transcript-stats</a> </span>

<div id="LR_EC_gene_table" class="scrollableTable"></div>

<br/><br/>
<h1>Transcript read alignment viewer</h1>
Delta value is the largest difference between the # of reads mapping to a gene in a tool and in raw reads.<br/>
<span style="color: red;">BE SURE TO DO THIS IF YOU WANT TO SEE THE IGV PLOTS: <a href="https://gitlab.com/leoisl/LR_EC_analyser#viewing-igv-plots-gene-stats-and-transcript-stats" target="_blank">https://gitlab.com/leoisl/LR_EC_analyser#viewing-igv-plots-gene-stats-and-transcript-stats</a> </span>

<div id="LR_EC_transcript_table" class="scrollableTable"></div>
<HTMLUncommentOutSimpleReport>


<br/><br/><br/><br/><br/><br/>
<h1>Appendix</h1>
<h2>Paralogous families:</h2>
<p>These are some data on the paralogous families given with the --paralogous parameter. <b>It has nothing to do with the raw reads or corrections.</b></p>
<p>We added this here so we can understand better what is being considered paralogous families.</p>
<p>Download the paralogous gene families here: <a href="lib/data/paralogous_families.txt", target="_blank">paralogous_families.txt</a></p>
<h3>Paralogous gene families size distribution: </h3>
<htmlParalogousGeneFamiliesSizeBarPlot>
</body>
</html>